crates/rust-analyzer/src/lsp_ext.rs

```diff
@@ -682,7 +682,9 @@ pub struct CompletionResolveData {
 }
 
 #[derive(Debug, Serialize, Deserialize)]
-pub struct InlayHintResolveData {}
+pub struct InlayHintResolveData {
+    pub file_id: u32,
```

Comment:
We should not expose internal file ids like this, this should be a `TextDocumentIdentifier`

Reply:
Can you elaborate the "expose" part? 

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#inlayHint has its resolve data defined as 
```
/**
 * A data entry field that is preserved on an inlay hint between
 * a `textDocument/inlayHint` and a `inlayHint/resolve` request.
 */
data?: LSPAny;
```

so clients cannot really know how to parse the JSON properly and only the source server knows more, so it's not exposed?

--------------

Another big reason I went on with that is that URI are absolute, for `inlay_hints.rs` example, I have `"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"` string returned.

If we start sending in _every_ hint such long lines, that kind of beats the purpose of trimming the JSON data via resolve requests.

Reply:
Hmm, the problem with using the id is that it might point to a deleted file, so we'd need to check that (which the tranlsation from uri to id already does), otherwise we'll panic if the client would give us a file deletion notification before we process a corresponding resolve.

In this case using the ID would be fine for now, since they are not managed by salsa like interned IDs are I guess

Reply:
That seems to be relatively simple to solve, if we add a method inside `Vfs` to check for existence, pushed the commit, thank you for explaining.

There was no such method before, so I am a bit concerned too now. 
Thinking of the alternative way, we could write more code and check whether the hint could be resolved — and only add `data` with log URIs into those files. Passing numbers around still looks more compelling to me though, hence trying it first.

Reply:
I would hope that clients are conservative with resolving given resolved lints are not needed for merely viewing them, in which case sending the paths should be fine. It's just that sending IDs might silently break if we decide to add some point change how they are allocated (currently they always map to a specific path, but maybe that changes in the future)

Reply:
Sorry for pushing more discussions on that front, but I'm still not convinced. If you're tired of this, maybe @\matklad  or any other person could be summoned to persuade me? 🙂 

> I would hope that clients are conservative with resolving given resolved lints are not needed for merely viewing them, in which case sending the paths should be fine.

While I agree with that, curious what do you see bad in the premature resolving scenario, related to `file_id`?
From client's perspective, inlay hints are very ephemeral: only file open and scroll could add anything to the existing hint collection; editing, [re]moving the file, /refresh and all other operations I can think of, invalidate hints and that happens inevitably fast.
Moreover, there's no indication in hints' LSP what exactly to invalidate, so we have to invalidate all.

In brief moments of adding something to hints, the user may use their mouse to hover and resolve hints (that's how VSCode and other few editors I am aware of do that), using received `file_id`.
So, it cannot really get that stale? 

> It's just that sending IDs might silently break if we decide to add some point change how they are allocated (currently they always map to a specific path, but maybe that changes in the future)

That is also something that I'm not fully sure how to understand.
I angle it differently: in both `textDocument/inlayHint` and `inlayHint/resolve` use the same code path, `inlay_hints.rs::inlay_hints` function that requires `file_id`.
I merely take the file_id from the hint request and put it into the resolve data to pass into the same `inlay_hints` function.
So if the first request succeeded, the 2nd one should too (esp. given the ephemeral nature of the requests, forcing them to happen relatively close to each other in time).

I do one extra thing between, to get hints' range — I can put that range into the resolve data.

--------------------

To illustrate my bloating concern, here's the json difference in the same place inside `inlay_hints.rs` mentioned in the PR message:

```
request:
{"jsonrpc":"2.0","id":37,"method":"textDocument/inlayHint","params":{"textDocument":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"},"range":{"start":{"line":438,"character":0},"end":{"line":477,"character":0}}}}
```

<details>
  <summary>json with `file_id`s, 4112 characters</summary>
  
```json
{"jsonrpc":"2.0","id":37,"result":[{"position":{"line":470,"character":1},"label":[{"value":"fn inlay_hints"}],"paddingLeft":true,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":445,"character":10},"label":[{"value":": "},{"value":"ProfileSpan"},{"value":""}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":445,"character":27},"label":[{"value":"label:"}],"kind":2,"paddingLeft":false,"paddingRight":true,"data":{"file_id":0}},{"position":{"line":446,"character":12},"label":[{"value":": "},{"value":"Semantics"},{"value":"<'_, "},{"value":"RootDatabase"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":447,"character":12},"label":[{"value":": "},{"value":"SourceFile"},{"value":""}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":448,"character":12},"label":[{"value":": &"},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":450,"character":15},"label":[{"value":": "},{"value":"Vec"},{"value":"<"},{"value":"InlayHint"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":452,"character":21},"label":[{"value":": "},{"value":"SemanticsScope"},{"value":"<'_>"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":452,"character":36},"label":[{"value":"node:"}],"kind":2,"paddingLeft":false,"paddingRight":true,"data":{"file_id":0}},{"position":{"line":453,"character":23},"label":[{"value":": "},{"value":"FamousDefs"},{"value":"<'_, '_>"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":456,"character":17},"label":[{"value":": impl FnMut("},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">)"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":456,"character":25},"label":[{"value":": "},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":456,"character":33},"label":[{"value":"hints:"}],"kind":2,"paddingLeft":false,"paddingRight":true,"data":{"file_id":0}},{"position":{"line":458,"character":22},"label":[{"value":": "},{"value":"TextRange"},{"value":""}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":460,"character":35},"label":[{"value":": "},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":462,"character":92},"label":[{"value":"impl "},{"value":"Iterator"},{"value":"<"},{"value":"Item"},{"value":" = "},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">>"}],"kind":1,"paddingLeft":true,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":461,"character":34},"label":[{"value":"impl "},{"value":"Iterator"},{"value":"<"},{"value":"Item"},{"value":" = "},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">>"},{"value":""}],"kind":1,"paddingLeft":true,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":460,"character":41},"label":[{"value":""},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">"}],"kind":1,"paddingLeft":true,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":462,"character":40},"label":" -> bool","kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":462,"character":39},"label":[{"value":": &"},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}},{"position":{"line":474,"character":33},"label":[{"value":": &&"},{"value":"Semantics"},{"value":"<'_, "},{"value":"RootDatabase"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_id":0}}]}

```
  
</details>

for the visible editor range alone, pretty much repeated on every consequent edit.

With this patch and all inlay hint resolve capabilities enabled, for the same example I observe quite a footprint reduction:

<details>
  <summary>json with file URIs, 5960 characters</summary>
  
```json
{"jsonrpc":"2.0","id":22,"result":[{"position":{"line":470,"character":1},"label":[{"value":"fn inlay_hints"}],"paddingLeft":true,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":445,"character":10},"label":[{"value":": "},{"value":"ProfileSpan"},{"value":""}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":445,"character":27},"label":[{"value":"label:"}],"kind":2,"paddingLeft":false,"paddingRight":true,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":446,"character":12},"label":[{"value":": "},{"value":"Semantics"},{"value":"<'_, "},{"value":"RootDatabase"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":447,"character":12},"label":[{"value":": "},{"value":"SourceFile"},{"value":""}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":448,"character":12},"label":[{"value":": &"},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":450,"character":15},"label":[{"value":": "},{"value":"Vec"},{"value":"<"},{"value":"InlayHint"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":452,"character":21},"label":[{"value":": "},{"value":"SemanticsScope"},{"value":"<'_>"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":452,"character":36},"label":[{"value":"node:"}],"kind":2,"paddingLeft":false,"paddingRight":true,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":453,"character":23},"label":[{"value":": "},{"value":"FamousDefs"},{"value":"<'_, '_>"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":456,"character":17},"label":[{"value":": impl FnMut("},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">)"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":456,"character":25},"label":[{"value":": "},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":456,"character":33},"label":[{"value":"hints:"}],"kind":2,"paddingLeft":false,"paddingRight":true,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":458,"character":22},"label":[{"value":": "},{"value":"TextRange"},{"value":""}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":460,"character":35},"label":[{"value":": "},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":462,"character":92},"label":[{"value":"impl "},{"value":"Iterator"},{"value":"<"},{"value":"Item"},{"value":" = "},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">>"}],"kind":1,"paddingLeft":true,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":461,"character":34},"label":[{"value":"impl "},{"value":"Iterator"},{"value":"<"},{"value":"Item"},{"value":" = "},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">>"},{"value":""}],"kind":1,"paddingLeft":true,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":460,"character":41},"label":[{"value":""},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">"}],"kind":1,"paddingLeft":true,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":462,"character":40},"label":" -> bool","kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":462,"character":39},"label":[{"value":": &"},{"value":"SyntaxNode"},{"value":"<"},{"value":"RustLanguage"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}},{"position":{"line":474,"character":33},"label":[{"value":": &&"},{"value":"Semantics"},{"value":"<'_, "},{"value":"RootDatabase"},{"value":">"}],"kind":1,"paddingLeft":false,"paddingRight":false,"data":{"file_uri":{"uri":"file:///Users/someonetoignore/work/rust-analyzer/crates/ide/src/inlay_hints.rs"}}}]}

```
  
</details>

So, for one edit (usually people type more + there's https://github.com/rust-lang/rust-analyzer/pull/15529 that makes r-a to double that number) in one file (we have to requery every visible file at least, maybe visible tabs' files), one visible range (editors could apply heuristics and query for more hints around the visible range to avoid excessive text movements on scroll after applying the hints), we pass 1848 extra symbols for a relatively non-nested project (it gets worse the deeper the project gets away from `/`).

Still better than 10803 characters for the current version, but `file_id` looks so much better.
It seems relatively simple to detect if we have anything to resolve at all, so I will try to reduce these URIs by omitting them for unresolvable hints, let's see the comparison then.

Reply:
For some reason I thought the file uris would only appear in the resolve request which is obviously not the case :man_facepalming: The reason for me saying things could silently break if we used IDs is the same reason as to why interned salsa IDs don't make it in the LSP layer. But anyways, it is unlikely for our file ids to change in a way that makes them unstable/cause them to change the path they represent so this should be fine. I was just a bit overly cautious here.

Reply:
![image](https://github.com/rust-lang/rust-analyzer/assets/2690773/9b102e33-aaf5-4813-bc90-08bff69d2324)

Unfortunately, I've picked a bad example for the proposed "nothing to remove => omit resolve data" heuristics.
As usual, when it's needed, it takes a while to be found, so instead, I have copied the first test from `fn_lifetime_fn` module and got a synthetic response to compare:

<details>
  <summary>part of `hints_lifetimes` test</summary>
  
```json
{"jsonrpc":"2.0","id":2300,"result":[{"position":{"line":383,"character":1},"label":[{"value":"mod trying"}],"paddingLeft":true,"paddingRight":false,"data":{"file_id":3662}},{"position":{"line":351,"character":30},"label":"'a","paddingLeft":false,"paddingRight":true},{"position":{"line":351,"character":22},"label":"'a","paddingLeft":false,"paddingRight":true},{"position":{"line":351,"character":17},"label":"<'a>","paddingLeft":false,"paddingRight":false},{"position":{"line":360,"character":30},"label":"'c","paddingLeft":false,"paddingRight":true},{"position":{"line":366,"character":30},"label":"'a","paddingLeft":false,"paddingRight":true},{"position":{"line":366,"character":31},"label":"'a","paddingLeft":false,"paddingRight":true},{"position":{"line":366,"character":34},"label":"'a","paddingLeft":false,"paddingRight":true},{"position":{"line":366,"character":22},"label":"'a","paddingLeft":false,"paddingRight":true},{"position":{"line":366,"character":17},"label":"<'a>","paddingLeft":false,"paddingRight":false},{"position":{"line":376,"character":28},"label":"'self","paddingLeft":false,"paddingRight":true},{"position":{"line":376,"character":18},"label":"'self","paddingLeft":false,"paddingRight":true},{"position":{"line":376,"character":16},"label":"<'self>","paddingLeft":false,"paddingRight":false},{"position":{"line":379,"character":36},"label":"'self","paddingLeft":false,"paddingRight":true},{"position":{"line":379,"character":18},"label":"'self","paddingLeft":false,"paddingRight":true},{"position":{"line":379,"character":28},"label":"'a","paddingLeft":false,"paddingRight":true},{"position":{"line":379,"character":16},"label":"<'self, 'a>","paddingLeft":false,"paddingRight":false}]}
```
  
</details>

So we do clean away resolve data and it looks like before, but have to add a lot of it nonetheless, hence I would still propose to keep `file_id` inside.
I see that you're not opposed so far, I'll mention in the comment this note too, please feel free to continue the discussion if some uncertainties are left.